### ISSUE TITLE
Why is migration adapter nullable?

### ISSUE THREAD

**Issue by martenb** [cakephp/phinx#2200]
I found this commit, where many methods changed their return type to nullable. https://github.com/cakephp/phinx/commit/f385bbb433cc1ab80b3a0b0a791c0a902775fc7a

For example `AbstractMigration::getAdapter()` returns `null` or `AdapterInterface`. 
Is that even possible, that migration does not have adapter?

With static analysis (like phpstan) i must check everytime, that adapeter is not null...
```php
assert($this->getAdapter() !== null);
$this->getAdapter()->...
```

**Comment by dereuromark** [comment]
This most likely is due to legacy code expectations and can be fixed in a new major (1.0).

**Comment by MasterOdin** [comment]
It's only really possible that `getAdapter()` returns `null` within the constructor as we don't pass in the adapter there. However, before we call the `change`, `up`, or `down` methods, we do `setAdapter` on the migration (and seed) so at that point, yeah, `getAdapter()` should always return something.

I think modifying the `getAdapter()` method to throw an exception if it's `null` should be fine to put in a minor release as while it's a "breaking change", the people it gets broken for would be those that are doing `->getAdapter()` in a custom adapter, and then having it always return `null` which seems so unlikely to probably not matter.
